File: CleanArchitecture.API/Controllers/UsersController.cs
using CleanArchitecture.Application.Commands;
using CleanArchitecture.Application.DTOs;
using MediatR;
using Microsoft.AspNetCore.Mvc;

namespace CleanArchitecture.API.Controllers;

[ApiController]
[Route("api/[controller]")]
public class UsersController : ControllerBase
{
    private readonly IMediator _mediator;

    public UsersController(IMediator mediator)
    {
        _mediator = mediator;
    }

    [HttpPost]
    public async Task<ActionResult<UserDto>> CreateUser(CreateUserCommand command)
    {
        var user = await _mediator.Send(command);
        return CreatedAtAction(nameof(CreateUser), user);
    }
}

File: FSharp/src/Api/Controllers/UsersController.fs
namespace CleanArchitecture.Api.Controllers

open System
open MediatR
open Microsoft.AspNetCore.Mvc
open CleanArchitecture.Domain.Common
open CleanArchitecture.Domain.User
open CleanArchitecture.Application.Users

[<ApiController>]
[<Route("api/[controller]")>]
type UsersController(mediator: IMediator) =
    inherit ControllerBase()

    [<HttpGet("{id:guid}")>]
    member _.GetUserById(id: Guid) =
        async {
            let entityId = EntityId id
            let! result = mediator.Send(GetUserByIdQuery entityId) |> Async.AwaitTask

            return
                match result with
                | Some user -> OkObjectResult user :> IActionResult
                | None -> NotFoundResult() :> IActionResult
        }
        |> Async.StartAsTask

    [<HttpGet>]
    member _.GetAllUsers() =
        async {
            let! users = mediator.Send(GetAllUsersQuery()) |> Async.AwaitTask
            return OkObjectResult users :> IActionResult
        }
        |> Async.StartAsTask

    [<HttpPost>]
    member _.CreateUser([<FromBody>] request: {| Email: string; Username: string |}) =
        async {
            let createRequest =
                { Email = Email request.Email
                  Username = Username request.Username }

            let! user = mediator.Send(CreateUserCommand createRequest) |> Async.AwaitTask
            return CreatedResult($"/api/users/{user.Id}", user) :> IActionResult
        }
        |> Async.StartAsTask
